numaprom/watcher.py

```diff
@@ -0,0 +1,77 @@
+import os
+import time
+
+from omegaconf import OmegaConf
+from watchdog.observers import Observer
+from numalogic.config import NumalogicConf
+from watchdog.events import FileSystemEventHandler
+
+from numaprom._constants import CONFIG_DIR, DEFAULT_CONFIG_DIR
+from numaprom import NumapromConf, get_logger
+
+_LOGGER = get_logger(__name__)
+
+config = {}
+
+
+def load_configs():
+    schema: NumapromConf = OmegaConf.structured(NumapromConf)
+
+    conf = OmegaConf.load(os.path.join(CONFIG_DIR, "config.yaml"))
+    app_configs = OmegaConf.merge(schema, conf).configs
+
+    conf = OmegaConf.load(os.path.join(DEFAULT_CONFIG_DIR, "config.yaml"))
+    default_configs = OmegaConf.merge(schema, conf).configs
+
+    conf = OmegaConf.load(os.path.join(DEFAULT_CONFIG_DIR, "numalogic_config.yaml"))
+    schema: NumalogicConf = OmegaConf.structured(NumalogicConf)
+    default_numalogic = OmegaConf.merge(schema, conf)
+
+    return app_configs, default_configs, default_numalogic
+
+
+def update_configs():
+    app_configs, default_configs, default_numalogic = load_configs()
+
+    config["app_configs"] = dict()
+    for _config in app_configs:
+        config["app_configs"][_config.namespace] = _config
+
+    config["default_configs"] = dict(map(lambda c: (c.namespace, c), default_configs))
+    config["default_numalogic"] = default_numalogic
+
+    _LOGGER.info("Successfully updated configs - %s", config)
+    return config
+
+
+class ConfigHandler(FileSystemEventHandler):
+    def on_any_event(self, event):
+        if event.event_type == "created" or event.event_type == "modified":
+            _file = os.path.basename(event.src_path)
+            _dir = os.path.basename(os.path.dirname(event.src_path))
+
+            _LOGGER.info("Watchdog received %s event - %s/%s", event.event_type, _dir, _file)
+            update_configs()
+
+
+class Watcher:
+    def __init__(self, directories=None, handler=FileSystemEventHandler()):
```

Comment:
a quick look into the WatchDog doc tells me that it uses `inotify` behind the scenes. `inotify` can miss events during buffer overruns (if I remember correctly). We might eventually need a reconciler to gracefully handle lost events if we see this as a major problem in production.

Reply:
I looked into this, WatchDog does handle the `inotify` missing events issue. It does the following to handle it. 

Recursive watching: automatically monitors all subdirectories of a watched directory, so it can detect changes even if `inotify` misses some events

Coalescing events: combine multiple events for the same file or directory into a single event, reducing the risk of missing events.

Event queueing: queues events internally, so even if the system is under high load and events are missed by `inotify`, they will eventually be processed by Watchdog.

Reply:
thank you